core/src/main/java/com/linecorp/armeria/server/file/AbstractHttpFile.java

```diff
@@ -0,0 +1,333 @@
+        }
+        return headers;
+    }
+
+    @Override
+    public final HttpResponse read(Executor fileReadExecutor, ByteBufAllocator alloc) {
+        requireNonNull(fileReadExecutor, "fileReadExecutor");
+        requireNonNull(alloc, "alloc");
+
+        try {
+            final HttpFileAttributes attrs = readAttributes();
+            final HttpHeaders headers = readHeaders(attrs);
+            if (headers == null) {
+                return null;
+            }
+
+            final long length = attrs.length();
+            if (length == 0) {
+                // No need to stream an empty file.
+                return HttpResponse.of(headers);
+            }
+
+            return doRead(headers, length, fileReadExecutor, alloc);
+        } catch (Exception e) {
+            return HttpResponse.ofFailure(e);
+        }
+    }
+
+    /**
+     * Returns a new {@link HttpResponse} which streams the content of the file which follows the specified
+     * {@link HttpHeaders}.
+     *
+     * @param headers the {@link HttpHeaders} of the response
+     * @param length the content length. The returned {@link HttpResponse} must stream only as many bytes as
+     *               this value.
+     * @param fileReadExecutor the {@link Executor} which should be used for performing a blocking file I/O
+     * @param alloc the {@link ByteBufAllocator} which should be used for allocating an input buffer
+     *
+     * @return the {@link HttpResponse}, or {@code null} if the file does not exist.
+     * @throws IOException if failed to open the file. Note that an I/O error which occurred during content
+     *                     streaming will be notified via the returned {@link HttpResponse}'s error
+     *                     notification mechanism.
+     */
+    @Nullable
+    protected abstract HttpResponse doRead(HttpHeaders headers, long length,
+                                           Executor fileReadExecutor,
+                                           ByteBufAllocator alloc) throws IOException;
+
+    @Override
+    public HttpService asService() {
+        return (ctx, req) -> {
+            final HttpMethod method = ctx.method();
+            if (method != HttpMethod.GET && method != HttpMethod.HEAD) {
+                return HttpResponse.of(HttpStatus.METHOD_NOT_ALLOWED);
+            }
+
+            final HttpFileAttributes attrs = readAttributes();
+            if (attrs == null) {
+                return HttpResponse.of(HttpStatus.NOT_FOUND);
+            }
+
+            // See https://tools.ietf.org/html/rfc7232#section-6 for more information
+            // about how conditional requests are handled.
+
+            // Handle 'if-none-match' header.
+            final HttpHeaders reqHeaders = req.headers();
+            final String etag = generateEntityTag(attrs);
+            final String ifNoneMatch = reqHeaders.get(HttpHeaderNames.IF_NONE_MATCH);
+            if (etag != null) {
+                if (ifNoneMatch != null) {
```

Comment:
nit: How about merging these two lines into one?

Reply:
We actually can't tell if entity tag exists or not without calling the `entityTagFunction.apply()` because it is allowed to return `null`. Let me merge the conditions into a single `if`, though.